File: Tests/TrainingCSVTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

final class TrainingCSVTests: XCTestCase {
    private let today = DayStamp(raw: 20_000)
    private func session(_ day: DayStamp, name: String = "Daily") -> AnalysisExport.Session {
        AnalysisExport.Session(day: day, startedAt: Date(timeIntervalSince1970: Double(day.raw) * 86_400),
                               routineName: name, plannedReps: 1, completedReps: 1,
                               reps: [RepSummary(peakKg: 12, avgKg: 10)], sessionsPerDayTarget: 2)
    }
    private func input(_ sessions: [AnalysisExport.Session], _ maxes: [AnalysisExport.MaxEntry] = []) -> AnalysisExport.Input {
        AnalysisExport.Input(sessions: sessions, maxes: maxes, today: today, generatedOn: today)
    }
    private func pullCSV(_ input: AnalysisExport.Input, scope: AnalysisExport.CSVScope = .recent) -> AnalysisExport.CSVDocument {
        AnalysisExport.csv(input, scope: scope, detail: .pulls)
    }
    // Numeric assertions use unquoted data rows. The quoting case is checked separately.
    private func records(_ document: AnalysisExport.CSVDocument, type: String) -> [[String: String]] {
        let lines = document.text.components(separatedBy: "\r\n")
        let keys = lines[0].components(separatedBy: ",")
        return lines.filter { $0.hasPrefix(type + ",") }.map {
            Dictionary(uniqueKeysWithValues: zip(keys, $0.components(separatedBy: ",")))
        }
    }
    func testRangeIsInclusiveAndAllHistoryRetainsOldPulls() {
        let data = input([session(today), session(today - 55), session(today - 56)])
        XCTAssertEqual(pullCSV(data).sessionCount, 2)
        XCTAssertEqual(pullCSV(data, scope: .all).pullCount, 3)
        XCTAssertEqual(records(pullCSV(data), type: "workout").first?["daily_target"], "2")
    }
    func testWorkoutUsesHistoricalMaxButExportsNoOtherWorkoutsOrMaxHistory() {
        let s = session(today)
        let old = AnalysisExport.MaxEntry(grip: GripSpec(), side: .left, kg: 30, day: today - 100,
                                         recordedAt: s.startedAt.addingTimeInterval(-8_640_000))
        let future = AnalysisExport.MaxEntry(grip: GripSpec(), side: .left, kg: 90, day: today,
                                            recordedAt: s.startedAt.addingTimeInterval(1))
        let csv = pullCSV(input([s], [future, old]), scope: .workout)
        XCTAssertEqual(csv.sessionCount, 1)
        XCTAssertEqual(csv.maxCount, 0)
        XCTAssertEqual(records(csv, type: "pull").first?["max_at_start_kg"], "30.0")
        XCTAssertFalse(csv.text.contains("90.0"))
        XCTAssertTrue(csv.filename.hasSuffix(".csv"))
    }
    func testSkippedAndTimerOnlyForceIsBlankButRealZeroSurvives() {
        var s = session(today)
        s.reps = [RepSummary(peakKg: 0), RepSummary(peakKg: 20, outcome: .skipped)]
        var rows = records(pullCSV(input([s])), type: "pull")
        XCTAssertEqual(rows[0]["peak_kg"], "0.0")
        XCTAssertEqual(rows[1]["peak_kg"], "")
        s.timing = .timerOnly
        rows = records(pullCSV(input([s])), type: "pull")
        XCTAssertTrue(rows.allSatisfy { $0["peak_kg"] == "" && $0["avg_kg"] == "" })
    }
    func testQuotedUnicodeNotesAndFormulaLikeNamesSurviveSafely() {
        var s = session(today, name: "=HYPERLINK(\"test\")")
        s.notes = "Flexion, café\n\"felt good\""
        let csv = pullCSV(input([s])).text
        XCTAssertTrue(csv.contains("\"'=HYPERLINK(\"\"test\"\")\""))
        XCTAssertTrue(csv.contains("\"Flexion, café\n\"\"felt good\"\"\""))
    }
    func testNewlineWithoutAnyOtherCSVPunctuationIsQuoted() {
        var s = session(today)
        for note in ["first\nsecond", "first\rsecond", "first\r\nsecond"] {
            s.notes = note
            XCTAssertTrue(pullCSV(input([s])).text.contains("\"" + note + "\""))
        }
    }
    func testCurlIdentityTargetAndOutcomeRemainExplicit() {
        var s = session(today)
        s.reps = [RepSummary(grip: GripSpec(position: .fingerCurl), targetLoKg: 8, targetHiKg: 12, outcome: .earlyRelease)]
        let pull = records(pullCSV(input([s])), type: "pull")[0]
        XCTAssertEqual(pull["position"], "fingerCurl")
        XCTAssertEqual(pull["fingers"], "IMRL")
        XCTAssertEqual(pull["target_low_kg"], "8.0")
        XCTAssertEqual(pull["target_high_kg"], "12.0")
        XCTAssertEqual(pull["outcome"], "earlyRelease")
    }

    func testPermanentKeysAndSelfContainedPullsAcrossRangesAndNewWorkouts() {
        let original = session(today - 1)
        let alone = records(pullCSV(input([original]), scope: .workout), type: "pull")[0]
        let expanded = records(pullCSV(input([session(today), original]), scope: .all), type: "pull")[1]
        XCTAssertEqual(alone["workout"], original.id.uuidString.lowercased())
        XCTAssertEqual(alone, expanded)
        XCTAssertEqual(alone["routine"], "Daily")
        XCTAssertEqual(alone["date"], AnalysisExport.isoDay(today - 1))
        XCTAssertEqual(alone["time_source"], "not_recorded")
    }

    func testSummaryPreservesOutcomesWeightedLoadsAndSeparateHandsAndPrescriptions() {
        var s = session(today)
        s.reps = [RepSummary(heldSeconds: 2, peakKg: 12, avgKg: 10),
                  RepSummary(heldSeconds: 6, peakKg: 24, avgKg: 20, outcome: .earlyRelease),
                  RepSummary(peakKg: 999, outcome: .skipped),
                  RepSummary(side: .right, heldSeconds: 3, peakKg: 9, avgKg: 8),
                  RepSummary(targetLoKg: 5, targetHiKg: 10)]
        let rows = records(AnalysisExport.csv(input([s])), type: "set")
        XCTAssertEqual(rows.count, 3)
        XCTAssertEqual(rows[0]["recorded"], "3")
        XCTAssertEqual(rows[0]["completed"], "1")
        XCTAssertEqual(rows[0]["held_s"], "8.0")
        XCTAssertEqual(rows[0]["peak_kg"], "24.0")
        XCTAssertEqual(rows[0]["avg_kg"], "17.5")
        XCTAssertEqual(rows[0]["outcome"], "completed:1|earlyRelease:1|skipped:1")
        XCTAssertEqual(rows[1]["hand"], "right")
        XCTAssertEqual(rows[2]["target_low_kg"], "5.0")
        s.timing = .timerOnly
        XCTAssertTrue(records(AnalysisExport.csv(input([s])), type: "set").allSatisfy { $0["peak_kg"] == "" && $0["avg_kg"] == "" })
    }

    func testElapsedTimingSavedRestAndBenchmarkProvenance() {
        var s = session(today)
        let plan = SessionPlan(sets: [SetPlan(repsPerSide: 2)], handMode: .bothHands,
                               holdSeconds: 10, restSeconds: 20, leadInSeconds: 3)
        s.plan = plan
        s.finishedAt = s.startedAt.addingTimeInterval(90)
        s.reps = [RepSummary(side: .both, startedElapsedSeconds: 3, endedElapsedSeconds: 13),
                  RepSummary(repIndex: 1, side: .both, startedElapsedSeconds: 43, endedElapsedSeconds: 53)]
        let maxEntry = AnalysisExport.MaxEntry(kg: 40, recordedAt: s.startedAt.addingTimeInterval(-1))
        var rows = records(pullCSV(input([s], [maxEntry])), type: "pull")
        XCTAssertEqual(rows[0]["planned_lead_in_s"], "3")
        XCTAssertEqual(rows[0]["planned_rest_s"], "20")
        XCTAssertEqual(rows[1]["gap_before_s"], "30.0")
        XCTAssertEqual(rows[0]["max_reference"], "both_hands")
        XCTAssertEqual(records(pullCSV(input([s])), type: "workout")[0]["time_source"], "runner_completion")
        s.reps[0].side = .left
        rows = records(pullCSV(input([s], [maxEntry])), type: "pull")
        XCTAssertEqual(rows[0]["max_reference"], "both_hands_fallback")
        XCTAssertEqual(records(pullCSV(input([s])), type: "pull")[0]["max_reference"], "no_recorded_max_at_start")
        s.reps[0].startedElapsedSeconds = -1
        XCTAssertEqual(records(pullCSV(input([s])), type: "pull")[0]["started_elapsed_s"], "")
    }

    func testSummaryDoesNotMergePrescriptionsThatRoundToTheSameDisplayValue() {
        var s = session(today)
        s.reps = [RepSummary(targetLoKg: 8.01, targetHiKg: 12), RepSummary(targetLoKg: 8.04, targetHiKg: 12)]
        XCTAssertEqual(records(AnalysisExport.csv(input([s])), type: "set").count, 2)
    }
}

File: android/engine/src/test/kotlin/run/nuri/getagrip/engine/TrainingCSVTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.engine

import java.time.Instant
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class TrainingCSVTests {
    private fun pullCSV(input: AnalysisExport.Input, scope: AnalysisExport.CSVScope = AnalysisExport.CSVScope.recent) =
        AnalysisExport.csv(input, scope, AnalysisExport.CSVDetail.pulls)
    private val today = DayStamp(20_000)
    private fun session(day: DayStamp = today) = AnalysisExport.Session(day = day,
        startedAt = Instant.ofEpochSecond(day.raw.toLong() * 86_400), routineName = "Daily",
        plannedReps = 1, completedReps = 1, reps = listOf(RepSummary(peakKg = 12.0, avgKg = 10.0)),
        sessionsPerDayTarget = 2)
    private fun input(sessions: List<AnalysisExport.Session>, maxes: List<AnalysisExport.MaxEntry> = emptyList()) =
        AnalysisExport.Input(sessions = sessions, maxes = maxes, today = today, generatedOn = today)
    private fun records(csv: AnalysisExport.CSVDocument, type: String): List<Map<String, String>> {
        val lines = csv.text.split("\r\n")
        val keys = lines[0].split(",")
        return lines.filter { it.startsWith("$type,") }.map { keys.zip(it.split(",")).toMap() }
    }
    @Test fun rangeIncludesTheFirstDayAndAllHistoryKeepsOldPulls() {
        val data = input(listOf(session(), session(today - 55), session(today - 56)))
        assertEquals(2, pullCSV(data).sessionCount)
        assertEquals(3, pullCSV(data, AnalysisExport.CSVScope.all).pullCount)
        assertEquals("2", records(pullCSV(data), "workout")[0]["daily_target"])
    }
    @Test fun singleWorkoutUsesOldMaxButNeverFutureMaxOrUnrelatedRecords() {
        val s = session()
        val old = AnalysisExport.MaxEntry(grip = GripSpec(), side = Side.left, kg = 30.0,
            day = today - 100, recordedAt = s.startedAt.minusSeconds(8_640_000))
        val future = old.copy(kg = 90.0, day = today, recordedAt = s.startedAt.plusSeconds(1))
        val csv = pullCSV(input(listOf(s), listOf(future, old)), AnalysisExport.CSVScope.workout)
        assertEquals(1, csv.sessionCount)
        assertEquals(0, csv.maxCount)
        assertEquals("30.0", records(csv, "pull")[0]["max_at_start_kg"])
        assertFalse(csv.text.contains("90.0"))
    }
    @Test fun unmeasuredForceIsBlankAndMeasuredZeroSurvives() {
        val s = session().copy(reps = listOf(RepSummary(peakKg = 0.0), RepSummary(peakKg = 20.0, outcome = RepOutcome.skipped)))
        val rows = records(pullCSV(input(listOf(s))), "pull")
        assertEquals("0.0", rows[0]["peak_kg"])
        assertEquals("", rows[1]["peak_kg"])
        assertTrue(records(pullCSV(input(listOf(s.copy(timing = AnalysisExport.Timing.timerOnly)))), "pull")
            .all { it["peak_kg"] == "" && it["avg_kg"] == "" })
    }
    @Test fun quotesUnicodeAndMultilineNotesArePreservedAndFormulasProtected() {
        val s = session().copy(routineName = "=HYPERLINK(\"test\")", notes = "Flexion, café\n\"felt good\"")
        val csv = pullCSV(input(listOf(s))).text
        assertTrue(csv.contains("\"'=HYPERLINK(\"\"test\"\")\""))
        assertTrue(csv.contains("\"Flexion, café\n\"\"felt good\"\"\""))
    }
    @Test fun curlAndFrozenPrescriptionRemainExplicit() {
        val s = session().copy(reps = listOf(RepSummary(grip = GripSpec(position = GripPosition.fingerCurl),
            targetLoKg = 8.0, targetHiKg = 12.0, outcome = RepOutcome.earlyRelease)))
        val pull = records(pullCSV(input(listOf(s))), "pull")[0]
        assertEquals("fingerCurl", pull["position"])
        assertEquals("IMRL", pull["fingers"])
        assertEquals("8.0", pull["target_low_kg"])
        assertEquals("12.0", pull["target_high_kg"])
        assertEquals("earlyRelease", pull["outcome"])
    }

    @Test fun stableKeysAndSelfContainedContextSurviveNewWorkouts() {
        val original = session(today - 1)
        val alone = records(pullCSV(input(listOf(original)), AnalysisExport.CSVScope.workout), "pull")[0]
        val expanded = records(pullCSV(input(listOf(session(), original)), AnalysisExport.CSVScope.all), "pull")[1]
        assertEquals(original.id.toString(), alone["workout"])
        assertEquals(alone, expanded)
        assertEquals("Daily", alone["routine"])
        assertEquals(AnalysisExport.isoDay(today - 1), alone["date"])
        assertEquals("not_recorded", alone["time_source"])
    }

    @Test fun summaryPreservesWeightedForcesOutcomesHandsAndPrescriptions() {
        val s = session().copy(reps = listOf(
            RepSummary(heldSeconds = 2.0, peakKg = 12.0, avgKg = 10.0),
            RepSummary(heldSeconds = 6.0, peakKg = 24.0, avgKg = 20.0, outcome = RepOutcome.earlyRelease),
            RepSummary(peakKg = 999.0, outcome = RepOutcome.skipped),
            RepSummary(side = Side.right, heldSeconds = 3.0, peakKg = 9.0, avgKg = 8.0),
            RepSummary(targetLoKg = 5.0, targetHiKg = 10.0)))
        val rows = records(AnalysisExport.csv(input(listOf(s))), "set")
        assertEquals(3, rows.size)
        assertEquals("3", rows[0]["recorded"])
        assertEquals("1", rows[0]["completed"])
        assertEquals("8.0", rows[0]["held_s"])
        assertEquals("24.0", rows[0]["peak_kg"])
        assertEquals("17.5", rows[0]["avg_kg"])
        assertEquals("completed:1|earlyRelease:1|skipped:1", rows[0]["outcome"])
        assertEquals("right", rows[1]["hand"])
        assertEquals("5.0", rows[2]["target_low_kg"])
    }

    @Test fun elapsedTimingSavedRestAndMaxProvenanceAreExplicit() {
        val s = session().copy(plan = SessionPlan(sets = listOf(SetPlan(repsPerSide = 2)), handMode = HandMode.bothHands,
            holdSeconds = 10, restSeconds = 20, leadInSeconds = 3),
            reps = listOf(RepSummary(side = Side.both, startedElapsedSeconds = 3.0, endedElapsedSeconds = 13.0),
                RepSummary(repIndex = 1, side = Side.both, startedElapsedSeconds = 43.0, endedElapsedSeconds = 53.0)))
        val max = AnalysisExport.MaxEntry(kg = 40.0, recordedAt = s.startedAt.minusSeconds(1))
        val rows = records(pullCSV(input(listOf(s), listOf(max))), "pull")
        assertEquals("3", rows[0]["planned_lead_in_s"])
        assertEquals("20", rows[0]["planned_rest_s"])
        assertEquals("30.0", rows[1]["gap_before_s"])
        assertEquals("both_hands", rows[0]["max_reference"])
        val left = s.copy(reps = listOf(s.reps.first().copy(side = Side.left)))
        assertEquals("both_hands_fallback", records(pullCSV(input(listOf(left), listOf(max))), "pull")[0]["max_reference"])
        assertEquals("no_recorded_max_at_start", records(pullCSV(input(listOf(left))), "pull")[0]["max_reference"])
    }

    @Test fun summaryDoesNotMergePrescriptionsThatRoundToTheSameDisplayValue() {
        val s = session().copy(reps = listOf(RepSummary(targetLoKg = 8.01, targetHiKg = 12.0), RepSummary(targetLoKg = 8.04, targetHiKg = 12.0)))
        assertEquals(2, records(AnalysisExport.csv(input(listOf(s))), "set").size)
    }
}
